### PR TITLE
LUCENE-9215: replace checkJavaDocs.py with doclet

### PR DESCRIPTION
**Pull request by rmuir**
This has the same logic as the previous python, but no longer relies
upon parsing HTML output, instead using java's doclet processor.

The errors are reported like "normal" javadoc errors with source file
name and line number and happen when running "gradlew javadoc"

Although the "rules" are the same as the previous python, the python had
some bugs where the checker didn't quite do exactly what we wanted, so
some fixes were applied throughout.

NOTE: I'm aware this isn't perfect, but I think it is much better than what we had before. 
The branch is pushed so if you want to help, push commits :)

### CODE REVIEW

**Comment by rmuir** [comment]
I ran a quick benchmark of just the javadoc task of `master` vs `LUCENE-9215` branch:
`./gradlew clean compileJava && time ./gradlew javadoc`

master: BUILD SUCCESSFUL in 3m 21s
LUCENE-9215: BUILD SUCCESSFUL in 3m 22s

So this check pass is effectively free and doesn't slow down javadoc processing at all (the hard part is already done), and I think it is a lot easier dealing with the errors directly from `gradlew javadoc`: things like filenames and line numbers really help. Plus we remove the previous python script which was recursively parsing HTML, and that thing was never instant.

**Comment by mocobeta** [comment]
I ran the javadoc task and saw how it works.
Would it be possible to pass the `javadocMissingLevel` property to the javadoc task as `Input` (as old `CheckMissingDocsTask` did so), so that we can force Gradle to re-run the checks when the property is changed without running `clean`?

**Comment by dweiss** [comment]
This looks really good already! My personal preference would be to keep the processing options inside the render javadoc file, rather than pushing them down to each project's build.gradle but it can definitely wait. Other than that I kept wondering if this should piggyback public/private scope changes to classes. Those weight classes in particular - making them private will have side-effects for other code (that may be doing an instanceof, for example).

**Comment by uschindler** [comment]
I agree with Dawid. At the beginning I did not like his approach to centralize everything, but this makes maintenance easier, as you have a better overview.
The build.gradle files per module are just metadata for Maven, nothing more. I kinda link this now.

**Comment by uschindler** [comment]
I would make the settings of the linter `@Input` options of the RenderJavadocs task, prepoulated with the project variables (like Javadocs URLs).

**Comment by dweiss** [comment]
Just to be clear, Robert - I can correct this, let me know if you're ok with it.

**Comment by rmuir** [comment]
Thanks @dweiss @uschindler @mocobeta  please, make any changes if you have the time and motiviation. I agree with the suggestions: plumbing the correct inputs sounds great to make it easier to fix the problems (no need to run clean). As far as how it is organized or how problems are fixed, it doesn't really matter to me, I was just trying to get things working. 

It is true i took a somewhat risky approach to make some things private. If someone really wants a class to be public they could at least have a one-liner describing what it does. So alternatively these classes could stay public but get some documentation (it means doing real work though)

I added the missing `@Override`s because I use the presence of the annotation as a proxy to determine that a method is overridden (and that java will be copied). Maybe it is not theoretically the best way, but it was simple and there was a lot to fix here. I know at least the ECJ compiler has the option to enforce that `@Override` is always there, maybe we should turn  that on too separately.

File: buildSrc/src/main/java/org/apache/lucene/gradle/MissingDoclet.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        return true;
+      }
+    }
+    return false;
+  }
+  
+  /** 
+   * Return true if the method is synthetic enum method (values/valueOf).
+   * According to the doctree documentation, the "included" set never includes synthetic elements.
+   * It should not happen but it happens!
```
[the remark this answers is not in the document]
Comment (rmuir) [reply buildSrc/src/main/java/org/apache/lucene/gradle/MissingDoclet.java:262]: yeah i forgot to attribute this to the author!

**Comment by uschindler** [comment]
> a318310 Made missing doclet a regular project. This also means all the checks are applied to it (forbidden APIs, etc.).

I was so happy that `lucene/tools` was gone. Why can't we keep it in buildSrc and just enable forbiddenapis and other checks there?

**Comment by dweiss** [comment]
It's a bit of a chicken and egg problem (boostrapping gradle itself). I do think a separate project for this is beneficial here - at least it's clear what it is, has a (potentially) separate set of dependencies, runs checks much like anything else, etc. We don't have to name it "tools"; it can be named more appropriately ('infrastructure/', 'internal/', 'build-tools/'). We also don't have to publish it as an artifact.

**Comment by uschindler** [comment]
Hi,
I still disagree. We have many classes in the build that are Groovy, based (like the javadocs task). Wht's the difference to that on, just because it's written in Java. If that's the case, I'd copypaste Robert's code into the render-javadocs.gradle file (as the syntax of Java is mostly the same with Gradle).

It also now produces Javadocs, it appears in documentation index, it produces artifacts,... It would really be easier if it would be part of the regular build infrastructure in build-src. I know from Elasticserach that they also run forbiddenapis checks there. Just declare and apply the plugin  and add one snippet to enable forbiddenapis for this one. I also don't like it to be not separated from the rest of the build. It's purely a build tool, so it should be in build-src or the grade subdirectory.

Last but not least: It should not be below "lucene", as it will be useful for Solr, too (after split of projects).

So please revert this commit, I disagree here. Maybe @rmuir has similar opinons.

**Comment by uschindler** [comment]
> If that's the case, I'd copypaste Robert's code into the render-javadocs.gradle file as another inner class (as the syntax of Java is mostly the same with Gradle).

OK, this won't work, as javadoc is running as external tool and we need a classpath. But nevertheless, it shouldn't be in Lucene.

Maybe we should do the same thing like with forbiddenapis:
- If Robert agrees, we put it as a simple Maven Project (no Gradle needed) into https://github.com/policeman-tools as a new project, Robert is also part of that organization.
- I can publish it on Maven using the Sonatype infrastructure afterwards:
- Everybody can use it as its a plain JAR file that can be added to any Project (Ant, Maven, Gradle, or maybe also Bazel or Makefiles that invoke Javadocs) so it could also be used by others who want to do missing docs checks in Maven.

**Comment by dweiss** [comment]
Making it a true project discovered two bugs the moment I tried to compile it. So there are benefits. It also makes development faster, at least in the early stage (because it doesn't invalidate caches). Sure we can externalize it as a separate dependency, I don't have a problem with this at all.

**Comment by dweiss** [comment]
A yet another alternative is to use a composite build to compile it locally. Then it is a separate project (top-level project) on its own and is not part of the Lucene/ Solr projects tree. This does take away some of the benefits I mentioned though.

**Comment by dweiss** [comment]
I did move it to a composite subproject. So it is now separate from Solr and Lucene, yet still in the source tree. We can move it back to buildSrc or anything else once it's ready and polished. For now, let's leave it as is - speeds up experimenting.

File: gradle/validation/forbidden-apis.gradle [hunk 2]
```diff
@@ -103,6 +103,13 @@ allprojects { prj ->
       ]
     }
 
+    // Doclet does use exotic JDK APIs.
```
Comment (uschindler) [inline gradle/validation/forbidden-apis.gradle:106]: No longer needed.

**Comment by rmuir** [comment]
seems the build fails because of `var` use? that check is broken: master is java 11

**Comment by uschindler** [comment]
> seems the build fails because of var use? that check is broken: master is java 11

That check was added to make merging easier. See the related issues about it. I don't care.

Personally I am not a fan of {{var}}, but I don't care. For this case, I would just change the code, unless we have a decission in master. It should not block this merge.

**Comment by uschindler** [comment]
As this class is not part of any code in branch_8x, I think validateSourcePatterns should exclude it for this check. Can be quickyl implemented.

I am dealing with validateSourcePatterns anyways, as it was not yet fully ported to Gradle (it causes a stupid warning due to duplicate task defs).

**Comment by rmuir** [comment]
>> That check was added to make merging easier. See the related issues about it. I don't care.

That is really wrong to do. If master is on java 11, then master is on java 11. If i had more energy i would heavy commit the fix to master just for emphasis.

Either we are on java 11 or we are stuck on java 8, not some hybrid mix in between.

**Comment by uschindler** [comment]
> That is really wrong to do. If master is on java 11, then master is on java 11. If i had more energy i would heavy commit the fix to master just for emphasis.

I agree with you, but that's a whole different story. For now I will just exclude this code from the validateSourcePatterns until we have a decission on this.

**Comment by mocobeta** [comment]
> That is really wrong to do. If master is on java 11, then master is on java 11. If i had more energy i would heavy commit the fix to master just for emphasis.
> Either we are on java 11 or we are stuck on java 8, not some hybrid mix in between.

+1 from me...

**Comment by uschindler** [comment]
> +1 from me...

same from me, but this should not be heavy-committed without issue. There was a lengthly discussion already (search JIRA and mailing list).

I added the workaround here, just needs to be cleaned up in a separate issue.

**Comment by rmuir** [comment]
I've added missing javadocs and reset visibility for the remaining classes, with one or two exceptions (and if you look at those, its clear the visibility was straight up wrong, because nothing else in the class (fields, methods) is visible.

Frankly, If we want to get this in and remove the python, now is the time I think. Otherwise let's decide we really like the python because that goddamn thing has been *way* too hard to get rid of. I'm just simply not going to be able to keep up with the new changes in master due to time constraints.

**Review by dweiss — APPROVED** [review]

**Pull request merged**